kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,624 @@
+            let mut scanner = LogReplayScanner::new(
+                commit_file,
+                json_client.clone(),
+                filter.clone(),
+                expression_evaluator.clone(),
+            );
+            scanner.phase_1()?;
+            scanner.resolve_dvs();
+            scanner.into_scan_batches()
+        })
+        .flatten_ok()
+        .map(|res| res?);
+    Ok(result)
+}
+
+fn get_add_transform_expr() -> Expression {
+    Expression::Struct(vec![
+        column_expr!("add.path"),
+        column_expr!("add.size"),
+        column_expr!("add.modificationTime"),
+        column_expr!("add.stats"),
+        column_expr!("add.deletionVector"),
+        Expression::Struct(vec![column_expr!("add.partitionValues")]),
+    ])
+}
+
+struct LogReplayScanner {
+    has_cdc_action: bool,
+    // A map from path to the deletion vector from the remove action. It is guaranteed that there
+    // is an add action with the same path in this commit
+    remove_dvs: HashMap<String, DvInfo>,
+    commit_file: ParsedLogPath,
+    json_client: Arc<dyn JsonHandler>,
+    timestamp: i64,
+    filter: Option<DataSkippingFilter>,
+    expression_evaluator: Arc<dyn ExpressionEvaluator>,
+    add_paths: HashSet<String>,
+}
+
+impl LogReplayScanner {
+    fn new(
+        commit_file: ParsedLogPath,
+        json_client: Arc<dyn JsonHandler>,
+        filter: Option<DataSkippingFilter>,
+        expression_evaluator: Arc<dyn ExpressionEvaluator>,
+    ) -> Self {
+        Self {
+            timestamp: commit_file.location.last_modified,
+            commit_file,
+            json_client,
+            filter,
+            has_cdc_action: Default::default(),
+            remove_dvs: Default::default(),
+            expression_evaluator,
+            add_paths: Default::default(),
+        }
+    }
+    fn phase_1(&mut self) -> DeltaResult<()> {
+        let schema = Phase1Visitor::schema()?;
+        let action_iter =
+            self.json_client
+                .read_json_files(&[self.commit_file.location.clone()], schema, None)?;
+
+        for actions in action_iter {
+            let actions = actions?;
+
+            //let actions_arrow: &ArrowEngineData =
+            //    actions.as_ref().any_ref().downcast_ref().unwrap();
+            //println!("actions: {:?}", actions_arrow.record_batch());
+            // apply data skipping to get back a selection vector for actions that passed skipping
+            // note: None implies all files passed data skipping.
+            let filter_vector = self
+                .filter
+                .as_ref()
+                .map(|filter| filter.apply(actions.as_ref()))
+                .transpose()?;
+
+            // we start our selection vector based on what was filtered. we will add to this vector
+            // below if a file has been removed
+            let selection_vector = match filter_vector {
+                Some(ref filter_vector) => filter_vector.clone(),
```

Comment:
Why do we need to clone it? I don't see any later uses...
```rust
            let selection_vector = self
                .filter
                .as_ref()
                .map(|filter| filter.apply(actions.as_ref()))
                .transpose()?
                .unwrap_or_else(|| vec![true, actions.len()]);
```
(again below -- maybe we should find a way to consolidate that very similar code?)

Reply:
Yeah, consolidating makes a lot of sense. This feels like it should be part of the DataSkippingFilter, since it's similar code to scan. What do you think?

Reply:
Are you suggesting to add an associated method that encapsulates the above logic, like this?
```rust
impl DataSkippingFilter {
    fn create_selection_vector(filter: &Option<Self>, data: &dyn EngineData) -> DeltaResult<Vec<bool>> { .. }
```
If so, that seems like a good idea to me!

Reply:
Yep! Shoulda made myself clearer lol. I can take that on in a followup PR.

Reply:
Update: We only do data skipping once during CDF replay now... but this still seems useful since CDF and normal replay both would use it?

Reply:
Yep both of them use it.  It'd certainly be nice to have

Reply:
follow-up issue?